akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -0,0 +1,433 @@
+  private[this] val maxQueue: Int = Math.min(props.queueSize, maxActivations)
+  @volatile private[this] var _systemQueue: LatestFirstSystemMessageList = SystemMessageList.LNil
+
+  protected def maySend: Boolean = !isTerminating
+  protected def isTerminating: Boolean = ActorCell.isTerminating(_status)
+  protected def setTerminating(): Unit = if (!ActorCell.isTerminating(_status)) unsafe.getAndAddInt(this, statusOffset, terminatingBit)
+  protected def setClosed(): Unit = if (!isClosed(_status)) unsafe.getAndAddInt(this, statusOffset, closedBit)
+
+  private def handleException: Catcher[Unit] = {
+    case e: InterruptedException ⇒
+      publish(Error(e, self.path.toString, getClass, "interrupted during message send"))
+      Thread.currentThread.interrupt()
+    case NonFatal(e) ⇒
+      publish(Error(e, self.path.toString, getClass, "swallowing exception during message send"))
+  }
+
+  def send(msg: T): Unit =
+    try {
+      val old = unsafe.getAndAddInt(this, statusOffset, 1)
+      val oldActivations = activations(old)
+      // this is not an off-by-one: #msgs is activations-1 if >0
+      if (oldActivations > maxQueue) {
+        if (Debug) println(s"[$thread] $self NOT enqueueing $msg at status $old ($oldActivations > $maxQueue)")
+        // cannot enqueue, need to give back activation token
+        unsafe.getAndAddInt(this, statusOffset, -1)
+        system.eventStream.publish(Dropped(msg, self))
+      } else if (ActorCell.isTerminating(old)) {
+        if (Debug) println(s"[$thread] $self NOT enqueueing $msg at status $old (is terminating)")
+        unsafe.getAndAddInt(this, statusOffset, -1)
+        system.deadLetters ! msg
+      } else {
+        if (Debug) println(s"[$thread] $self enqueueing $msg at status $old")
+        // need to enqueue; if the actor sees the token but not the message, it will reschedule
+        queue.add(msg)
+        if (oldActivations == 0) {
+          if (Debug) println(s"[$thread] $self being woken up")
+          unsafe.getAndAddInt(this, statusOffset, 1) // the first 1 was just the “active” bit, now add 1msg
+          // if the actor was not yet running, set it in motion; spurious wakeups don’t hurt
+          executionContext.execute(this)
+        }
+      }
+    } catch handleException
+
+  def sendSystem(signal: SystemMessage): Unit = {
+    @tailrec def needToActivate(): Boolean = {
+      val currentList = _systemQueue
+      if (currentList.head == NoMessage) {
+        system.deadLetters.toImpl.sendSystem(signal)
+        false
+      } else {
+        unsafe.compareAndSwapObject(this, systemQueueOffset, currentList.head, (signal :: currentList).head) || {
+          signal.unlink()
+          needToActivate()
+        }
+      }
+    }
+    try {
+      if (needToActivate()) {
+        val old = unsafe.getAndAddInt(this, statusOffset, 1)
+        if (isClosed(old)) {
+          // nothing to do
+          if (Debug) println(s"[$thread] $self NOT enqueueing $signal: terminating")
```

Comment:
shouldn't this decrease by one? I mean, the unlikely case that these accumulate and overflow into the upper bits...